File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/lib/Archive/TrpakTypes.ts
/**
 * TRPAK Archive Type Definitions
 * Converted from C# TrpakTypes.cs
 */

import * as fs from 'fs';

// ============================================================================
// Binary Structs
// ============================================================================

export enum TrpakCompressionType {
    None = 0,
    Zlib = 1,
    Lz4 = 2,
    Oodle = 3
}

export interface TrpakHeader {
    magic: bigint;
    Version: number;
    Relocated: number;
    FileNumber: number;
    FolderNumber: number;
}

export namespace TrpakHeader {
    export const Magic = 0x4B434150_584C4647n; // "GFLXPACK"
    export const Size = 0x18;
}

export interface TrpakFolderHeader {
    Hash: bigint;
    ContentNumber: number;
    Reserved: number;
}

export namespace TrpakFolderHeader {
    export const Size = 0x10;
}

export interface TrpakFolderIndex {
    Hash: bigint;
    Index: number;
    Reserved: number;
}

export namespace TrpakFolderIndex {
    export const Size = 0x10;
}

export interface TrpakFileHeader {
    Level: number;
    CompressionType: TrpakCompressionType;
    BufferSize: number;
    FileSize: number;
    Reserved: number;
    FilePointer: bigint;
}

export namespace TrpakFileHeader {
    export const Size = 0x18;
}

// ============================================================================
// FNV Hash
// ============================================================================

/**
 * FNV-1a 64-bit hash used by TRPAK for file/folder name lookup.
 * Ported from gftool GFFNV.cs.
 */
export class FnvHash {
    private static readonly FnvPrime = 0x00000100000001b3n;
    private static readonly FnvBasis = 0xCBF29CE484222645n;

    static Hash(str: string): bigint {
        const buf = Buffer.from(str, 'utf8');
        let result = this.FnvBasis;
        for (const b of buf) {
            result ^= BigInt(b);
            result *= this.FnvPrime;
        }
        return result;
    }
}

// ============================================================================
// Hash Cache
// ============================================================================

/**
 * Maps FNV hashes back to human-readable file paths.
 * Ported from gftool GFPakHashCache.cs.
 */
export class TrpakHashCache {
    private _cache: Map<bigint, string> = new Map();

    get count(): number {
        return this._cache.size;
    }

    /**
     * Load a binary hash cache file (GFPAKHashCache.bin format).
     */
    LoadBinaryCache(path: string): void {
        if (!fs.existsSync(path)) return;
        const data = fs.readFileSync(path);
        let offset = 0;
        
        const count = data.readBigUInt64LE(offset);
        offset += 8;
        
        for (let i = 0n; i < count; i++) {
            const hash = data.readBigUInt64LE(offset);
            offset += 8;
            
            // Read .NET BinaryReader string format (7-bit encoded length prefix)
            let length = 0;
            let shift = 0;
            while (true) {
                const b = data[offset++];
                length |= (b & 0x7F) << shift;
                if ((b & 0x80) === 0) break;
                shift += 7;
            }
            
            const name = data.toString('utf8', offset, offset + length);
            offset += length;
            
            this._cache.set(hash, name);
        }
    }

    /**
     * Load a text hash list (one "hash path" per line).
     */
    LoadHashList(lines: string[]): void {
        for (const line of lines) {
            if (!line || line.trim().length === 0) continue;
            const parts = line.trim().split(/\s+/).filter(p => p.length > 0);
            if (parts.length < 2) continue;

            const parseResult = this.TryParseHex(parts[0]);
            if (parseResult.success) {
                this._cache.set(parseResult.value, parts[1]);
            } else {
                this._cache.set(FnvHash.Hash(parts[1]), parts[1]);
            }
        }
    }

    Add(hash: bigint, name: string): void {
        this._cache.set(hash, name);
    }

    GetName(hash: bigint): string | null {
        return this._cache.get(hash) ?? null;
    }

    private TryParseHex(text: string): { success: boolean; value: bigint } {
        if (!text || text.trim().length === 0) {
            return { success: false, value: 0n };
        }
        
        let trimmed = text.trim();
        if (trimmed.toLowerCase().startsWith('0x')) {
            trimmed = trimmed.slice(2);
        }
        
        try {
            const value = BigInt('0x' + trimmed);
            return { success: true, value };
        } catch {
            return { success: false, value: 0n };
        }
    }
}

// ============================================================================
// Output Models
// ============================================================================

export class TrpakArchive {
    Folders: TrpakFolder[] = [];

    /**
     * Find a file by its full path (case-insensitive).
     */
    FindFile(fullPath: string): TrpakFile | null {
        for (const folder of this.Folders) {
            for (const file of folder.Files) {
                if (file.FullName.toLowerCase() === fullPath.toLowerCase()) {
                    return file;
                }
            }
        }
        return null;
    }

    /**
     * Find all files matching a predicate.
     */
    FindFiles(predicate: (file: TrpakFile) => boolean): TrpakFile[] {
        const results: TrpakFile[] = [];
        for (const folder of this.Folders) {
            for (const file of folder.Files) {
                if (predicate(file)) {
                    results.push(file);
                }
            }
        }
        return results;
    }

    /**
     * Find all files with a specific extension.
     */
    FindFilesByExtension(extension: string): TrpakFile[] {
        return this.FindFiles(f => 
            f.FullName.toLowerCase().endsWith(extension.toLowerCase())
        );
    }
}

export class TrpakFolder {
    Path: string = '';
    Files: TrpakFile[] = [];
}

export class TrpakFile {
    Name: string = '';
    FullName: string = '';
    Data: Buffer = Buffer.alloc(0);
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Trpak/Archive/TrpakTypes.cs
using System.Runtime.InteropServices;
using System.Text;

namespace MiniToolbox.Trpak.Archive
{
    #region Binary Structs

    internal enum TrpakCompressionType : ushort
    {
        None = 0,
        Zlib = 1,
        Lz4 = 2,
        Oodle = 3
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct TrpakHeader
    {
        public const ulong Magic = 0x4B434150_584C4647; // "GFLXPACK"
        public const int Size = 0x18;
        public ulong magic;
        public uint Version;
        public uint Relocated;
        public uint FileNumber;
        public uint FolderNumber;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct TrpakFolderHeader
    {
        public const int Size = 0x10;
        public ulong Hash;
        public uint ContentNumber;
        public uint Reserved;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct TrpakFolderIndex
    {
        public const int Size = 0x10;
        public ulong Hash;
        public uint Index;
        public uint Reserved;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct TrpakFileHeader
    {
        public const int Size = 0x18;
        public ushort Level;
        public TrpakCompressionType CompressionType;
        public uint BufferSize;
        public uint FileSize;
        public uint Reserved;
        public ulong FilePointer;
    }

    #endregion

    #region FNV Hash

    /// <summary>
    /// FNV-1a 64-bit hash used by TRPAK for file/folder name lookup.
    /// Ported from gftool GFFNV.cs.
    /// </summary>
    public static class FnvHash
    {
        private const ulong FnvPrime = 0x00000100000001b3;
        private const ulong FnvBasis = 0xCBF29CE484222645;

        public static ulong Hash(string str)
        {
            byte[] buf = Encoding.UTF8.GetBytes(str);
            ulong result = FnvBasis;
            foreach (byte b in buf)
            {
                result ^= b;
                result *= FnvPrime;
            }
            return result;
        }
    }

    #endregion

    #region Hash Cache

    /// <summary>
    /// Maps FNV hashes back to human-readable file paths.
    /// Ported from gftool GFPakHashCache.cs.
    /// </summary>
    public class TrpakHashCache
    {
        private readonly Dictionary<ulong, string> _cache = new();

        public int Count => _cache.Count;

        /// <summary>
        /// Load a binary hash cache file (GFPAKHashCache.bin format).
        /// </summary>
        public void LoadBinaryCache(string path)
        {
            if (!File.Exists(path)) return;
            using var br = new BinaryReader(File.OpenRead(path));
            var count = br.ReadUInt64();
            for (ulong i = 0; i < count; i++)
            {
                var hash = br.ReadUInt64();
                var name = br.ReadString();
                _cache[hash] = name;
            }
        }

        /// <summary>
        /// Load a text hash list (one "hash path" per line).
        /// </summary>
        public void LoadHashList(IEnumerable<string> lines)
        {
            foreach (var line in lines)
            {
                if (string.IsNullOrWhiteSpace(line)) continue;
                var parts = line.Trim().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length < 2) continue;

                if (TryParseHex(parts[0], out ulong hash))
                    _cache[hash] = parts[1];
                else
                    _cache[FnvHash.Hash(parts[1])] = parts[1];
            }
        }

        public void Add(ulong hash, string name) => _cache[hash] = name;

        public void MergeFrom(Dictionary<ulong, string> entries)
        {
            foreach (var (hash, name) in entries)
                _cache.TryAdd(hash, name);
        }

        public string? GetName(ulong hash) => _cache.TryGetValue(hash, out var name) ? name : null;

        public IReadOnlyList<string> AllPaths() => _cache.Values.Distinct().ToList();

        private static bool TryParseHex(string text, out ulong value)
        {
            value = 0;
            if (string.IsNullOrWhiteSpace(text)) return false;
            text = text.Trim();
            if (text.StartsWith("0x", StringComparison.OrdinalIgnoreCase)) text = text[2..];
            return ulong.TryParse(text, System.Globalization.NumberStyles.HexNumber,
                System.Globalization.CultureInfo.InvariantCulture, out value);
        }
    }

    #endregion

    #region Output Models

    public class TrpakArchive
    {
        public List<TrpakFolder> Folders { get; } = new();

        /// <summary>
        /// Find a file by its full path (case-insensitive).
        /// </summary>
        public TrpakFile? FindFile(string fullPath)
        {
            foreach (var folder in Folders)
                foreach (var file in folder.Files)
                    if (string.Equals(file.FullName, fullPath, StringComparison.OrdinalIgnoreCase))
                        return file;
            return null;
        }

        /// <summary>
        /// Find all files matching a predicate.
        /// </summary>
        public IEnumerable<TrpakFile> FindFiles(Func<TrpakFile, bool> predicate)
        {
            return Folders.SelectMany(f => f.Files).Where(predicate);
        }

        /// <summary>
        /// Find all files with a specific extension.
        /// </summary>
        public IEnumerable<TrpakFile> FindFilesByExtension(string extension)
        {
            return FindFiles(f => f.FullName.EndsWith(extension, StringComparison.OrdinalIgnoreCase));
        }
    }

    public class TrpakFolder
    {
        public string Path { get; set; } = string.Empty;
        public List<TrpakFile> Files { get; } = new();
    }

    public class TrpakFile
    {
        public string Name { get; set; } = string.Empty;
        public string FullName { get; set; } = string.Empty;
        public byte[] Data { get; set; } = Array.Empty<byte>();
    }

    #endregion
}
